corehq/apps/registration/static/registration/js/user_login_form.js

```diff
@@ -118,8 +120,12 @@ hqDefine('registration/js/user_login_form', [
             self.proceedToNextStep();
         };
 
-        self.continueToSsoLogin = function (sso_url) {
-            window.location = sso_url;
+        self.continueToSsoLogin = function (ssoUrl) {
+            if (self.nextUrl) {
+                // note ssoUrl already contains ?username=foo
+                ssoUrl = ssoUrl + "&next=" + self.nextUrl;
+            }
+            window.location = ssoUrl;
```

Comment:
```suggestion
            ssoUrl = new URL(ssoUrl);
            if (self.nextUrl) {
                ssoUrl.set("next", self.nextUrl);
            }
            window.location = ssoUrl.toString();
```

Reply:
Not 100% sure if `.toString()` is required there. My JS is a bit weak.

Reply:
This doesn't work as there is no `set` method for `URL`. see https://developer.mozilla.org/en-US/docs/Web/API/URL/URL

Django / the one login saml library is the only thing that interacts with `next` and it checks to verify that "next" is in fact a valid URL see: https://github.com/onelogin/python3-saml/blob/master/src/onelogin/saml2/errors.py#L31.

Javascript doesn't execute this line at all treats it as a string because of how it is fetched due to `URLSearchParams`. I even tried passing `next=alert("hi");` to verify there was no javascript code injection going on. 

We need to pass in the value of `next` exactly as it is written in the URL `&next=/path/to/view' and not escape the slashes otherwise it messes up `OneLogin`'s RedirectURL and throws an error.

I'm not understanding what the problem is here?

Reply:
I didn't realize the limitations of the `URL` object in this context. I guess I was looking at some specific version that's browser-specific.

> We need to pass in the value of `next` exactly as it is written in the URL `&next=/path/to/view` and not escape the slashes

Slashes don't get escaped in query-string quoting, but other characters do (e.g. `&`, `#`, `?`, etc).

> I'm not understanding what the problem is here?

My concern is in regards to *"where exactly does the value of the `next` URL come from, and what guarantees do we have about its contents?"*.  Unless the application that first wrote that parameter has quoted the value twice, then the result of `urlParams.get('next')` will be the exact literal value. Perhaps I'm missing why it's ok to assume that that value is already query-string-safe?

For example, if the `next` value is `/foo&bar`, the resulting URL will be `...&next=/foo&bar`, which will become a problem when that query string is parsed.

Reply:
A yeah, django is typically the only thing pre-populating `next`, which is done when a user who has logged out visits a URL directly that they only have access to if logged in. Users are not filling in the next parameter manually anywhere...so this functionality basically forwards whatever the value is in `next` that Django left and passes it to the SSO login view.